### PR TITLE
cluster: Fix join shuffling and FloatingIP discrepancies

### PR DESCRIPTION
**Pull request by Baisang**
Before, it was possible for `machine.Machine`s to flip-flop between `db.Machine`s
in the cluster join. The join has been updated to now break ties between machines
by prioritizing joining machines that have the same PublicIP and other attributes.

Additionally, assigning FloatingIPs was updated. FloatingIPs are now only assigned
once the CloudID of a machine has been assigned. This ensures that FloatingIPs will
only be assigned once the machine has fully booted and been stabilized in the join
(meaning, it will not join with any other machine).

Tests have been corrected and added to reflect these changes.

==============================================================

Seemed fine locally with the logs test, but I haven't really been able to test this with floating IP outside of unit tests, so let me know if anything seems a little shaky.

### CODE REVIEW

**Comment by Baisang** [comment]
lol forgot to lint, one sec

File: cluster/cluster.go [hunk 1]
```diff
@@ -356,13 +357,12 @@ type syncDBResult struct {
...
 func syncDB(cms []machine.Machine, dbms []db.Machine) syncDBResult {
 	ret := syncDBResult{}
-	cms = foreman.GetMachineRoles(cms)
 
 	pair1, dbmis, cmis := join.Join(dbms, cms, func(l, r interface{}) int {
 		dbm := l.(db.Machine)
 		m := r.(machine.Machine)
 
-		// Don't join if m.ID and dbm.CloudID are empty.
+		// Don't join if m.ID is empty (unit tests)
```
Comment (ejj) [inline cluster/cluster.go]: Why this (unit tests) note?

File: cluster/cluster.go [hunk 2]
```diff
@@ -378,25 +378,29 @@ func syncDB(cms []machine.Machine, dbms []db.Machine) syncDBResult {
...
 			dbm.Region != m.Region ||
 			dbm.Size != m.Size ||
 			dbm.Preemptible != m.Preemptible ||
 			(m.DiskSize != 0 && dbm.DiskSize != m.DiskSize) ||
-			(m.Role != db.None && dbm.Role != m.Role):
+			(m.Role != db.None && dbm.Role != m.Role) {
 			return -1
-		case m.ID != "" && dbm.CloudID == m.ID:
+		}
+		if m.ID != "" && dbm.CloudID == m.ID {
```
Comment (ejj) [inline cluster/cluster.go]: Why do we need this case at all then?

File: cluster/cluster.go [hunk 3]
```diff
@@ -378,25 +378,29 @@ func syncDB(cms []machine.Machine, dbms []db.Machine) syncDBResult {
...
 		}
+
+		score := 10
+		if dbm.Role != db.None && m.Role != db.None && dbm.Role == m.Role {
+			score -= 4
+		}
+		if dbm.FloatingIP == m.FloatingIP {
+			score -= 2
+		}
+		if dbm.PublicIP == m.PublicIP && dbm.PrivateIP == m.PrivateIP {
```
Comment (ejj) [inline cluster/cluster.go]: Personally, I think that the IP being correct should be a bigger weight than the floating IP being correct.  That way you'd be preferring fundamental machine attributes over things that can be moved to that machine

**Comment by ejj** [comment]
LGTM, just a couple of little nits

File: cluster/cluster.go (hunk 3 above)
Comment (kklin) [inline cluster/cluster.go]: +1

File: cluster/cluster.go (hunk 1 above)
Comment (Baisang) [inline cluster/cluster.go]: `m.ID` shouldn't really be empty in real life, it only is for some unit tests. The alternative to this is correcting all of the unit tests that lack `m.ID` currently, I'm not sure what is best

File: cluster/cluster.go (hunk 2 above)
Comment (Baisang) [inline cluster/cluster.go]: I left it in from previous, we can remove it if you think not necessary

File: cluster/cluster.go (hunk 3 above)
Comment (Baisang) [inline cluster/cluster.go]: Makes sense, I'll try it out

File: cluster/cluster.go (hunk 2 above)
Comment (ejj) [inline cluster/cluster.go]: Doesn't look necessary to me, but I'd like you to convince yourself either way =)

**Comment by Baisang** [comment]
@ejj ready for review again

**Pull request merged**